### PR TITLE
feat: Section Ads

### PR DESCRIPTION
**Pull request by Geoff-Ford**
https://nidigitalsolutions.jira.com/browse/TNLT-3682

### CODE REVIEW

File: packages/section/__tests__/unit/utils.test.js [hunk 1]
```diff
@@ -63,7 +64,10 @@ describe("composeSliceBuilders", () => {
       { id: "l", name: "DailyUniversalRegister" },
     ];
 
-    const buildData = composeSliceBuilders(builderOne, builderTwo);
+    const buildData = prepareSlicesForRender(false, variants)(
+      builderOne,
```
Comment (aronshamash) [inline packages/section/__tests__/unit/utils.test.js:68]: I take it all these `builders` will go ?

File: packages/ad/src/utils/ad-init.js [hunk 2]
```diff
@@ -231,6 +231,11 @@ export default ({ el, data, platform, eventCallback, window }) => {
               const testGroup = ["a", "b", "c"].indexOf(slotName.slice(-1)) + 1;
               slot.setTargeting("testgroup", testGroup.toString());
             }
+            if (/^native-section-ad-/.test(slotName)) {
+              const testGroup = ["a", "b", "c"].indexOf(slotName.slice(-1)) + 1;
```
Comment (aronshamash) [inline packages/ad/src/utils/ad-init.js:235]: can this be DRYed up ?

File: packages/ad/src/utils/generate-config.js [hunk 3]
```diff
@@ -43,6 +43,8 @@ const sizeMap = {
   "native-inline-ad-a": sizes.native,
   "native-inline-ad-b": sizes.nativeB,
   "native-inline-ad-c": sizes.nativeC,
+  "native-section-ad-b": sizes.sectionB,
```
Comment (aronshamash) [inline packages/ad/src/utils/generate-config.js:46]: why no `sectionA`?

File: packages/pages/src/section/section.js [hunk 4]
```diff
@@ -131,9 +134,18 @@ class SectionPage extends Component {
...
+    const { error, isLoading, publicationName, variants } = this.props;
     const { recentlyOpenedPuzzleCount, savedArticles, section } = this.state;
 
+    const adConfig =
+      isLoading || error
+        ? {}
+        : adTargetConfig({
+            // adTestMode,
+            // article,
+            sectionName: section.name,
```
Comment (aronshamash) [inline packages/pages/src/section/section.js:146]: do we need a default or will this always be populated ?

File: packages/pages/src/section/ad-targeting-config.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import { NativeModules, Platform } from "react-native";
+
+const config = NativeModules.ReactConfig;
+
+export default ({ adTestMode = null, sectionName }) => ({
```
Comment (aronshamash) [inline packages/pages/src/section/ad-targeting-config.js:5]: do you think it is worth having a base ad config with all the common stuff in ?

File: packages/ad/src/utils/generate-config.js (hunk 3 above)
Comment (aronshamash) [reply packages/ad/src/utils/generate-config.js:46]: I worked it out, there is no need for it!

File: packages/variant-testing/src/variant-testing.tsx [hunk 6]
```diff
@@ -1,42 +1,53 @@
 import React, { createContext, useContext, ReactNode } from "react";
 import PropTypes from "prop-types";
 
+import { useResponsiveContext } from "@times-components-native/responsive";
+
 export const VariantTestingContext = createContext({});
 
+const validateVariant = (variant: string) => {
```
Comment (aronshamash) [inline packages/variant-testing/src/variant-testing.tsx:8]: yes better.

File: packages/ad/src/utils/ad-init.js (hunk 2 above)
Comment (c-j-j) [reply packages/ad/src/utils/ad-init.js:233]: You may have made this comment pre-DRYing - pushed a change to extract that into one regex

File: packages/pages/src/section/ad-targeting-config.js (hunk 5 above)
Comment (c-j-j) [reply packages/pages/src/section/ad-targeting-config.js:5]: Yes that will make sense - although  they are due to audit this config for section and articles, so exactly config should be sent for article vs section is unclear right now

File: packages/pages/src/section/section.js (hunk 4 above)
Comment (c-j-j) [reply packages/pages/src/section/section.js:146]: I had assumed section.name is always set - can we assume that's the case, and change if we see undefined section names?

File: packages/ad/src/utils/ad-init.js (hunk 2 above)
Comment (aronshamash) [reply packages/ad/src/utils/ad-init.js:233]: I did

File: packages/pages/src/section/section.js (hunk 4 above)
Comment (aronshamash) [reply packages/pages/src/section/section.js:146]: yes. we are doing `sectionName.toLowerCase()` so that would break if sectionName is `undefined`

**Review by aronshamash — APPROVED** [review]

File: packages/section/__tests__/unit/utils.test.js (hunk 1 above)
Comment (c-j-j) [reply packages/section/__tests__/unit/utils.test.js:68]: I think these tests can probably all go - they are trying to test the piping logic which is now done by the `pipe` util and its tested separately. They don't check any of the data transformation. 

I think I'll add some tests that check prepareSlicesForRender with real input/output data and snapshot the result - as t he internal bits of logic are also tested separately

**Pull request merged**